### PR TITLE
Add out-of-the-box health indicator for reactive MongoDB

### PR DESCRIPTION
**Pull request by qinnnyul**
Add out-of-the-box health indicator for reactive MongoDB and write few test cases to cover new changes. #11766

### CODE REVIEW

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/mongo/MongoReactiveDataAutoConfiguration.java [hunk 1]
```diff
@@ -73,4 +80,19 @@ public ReactiveMongoTemplate reactiveMongoTemplate(
 		return new ReactiveMongoTemplate(reactiveMongoDatabaseFactory, converter);
 	}
 
+	@Bean
+	@ConditionalOnMissingBean(MongoConverter.class)
+	public MappingMongoConverter mappingMongoConverter() {
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/mongo/MongoReactiveDataAutoConfiguration.java]: How is this related to this change?

File: spring-boot-project/spring-boot-actuator/src/test/java/org/springframework/boot/actuate/mongo/MongoReactiveHealthIndicatorTest.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+import static org.mockito.Mockito.mock;
+
+/**
+ * Tests for {@link MongoReactiveHealthIndicator}.
+ *
+ * @author Yulin Qin
+ */
+
+public class MongoReactiveHealthIndicatorTest {
+	@Mock
```
Comment (snicoll) [inline spring-boot-project/spring-boot-actuator/src/test/java/org/springframework/boot/actuate/mongo/MongoReactiveHealthIndicatorTest.java]: We don't use this construct, please mock the template in the test as you've done for the rest.

File: spring-boot-project/spring-boot-actuator-autoconfigure/pom.xml [hunk 3]
```diff
@@ -283,6 +283,16 @@
 			<artifactId>spring-data-mongodb</artifactId>
 			<optional>true</optional>
 		</dependency>
+		<dependency>
```
Comment (snicoll) [inline spring-boot-project/spring-boot-actuator-autoconfigure/pom.xml]: Dependencies are ordered alphabetically. Please reorder those two entries.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/mongo/MongoReactiveDataAutoConfiguration.java (hunk 1 above)
Comment (qinnnyul) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/mongo/MongoReactiveDataAutoConfiguration.java]: It is required by “ReactiveMongoTemplate” bean while it is missing from Spring IOC factory, that is why I provided default implementation of MongoConverter here when it is missing in context, otherwise, it will complains “unsatisfied dependencies ”

File: spring-boot-project/spring-boot-actuator/src/test/java/org/springframework/boot/actuate/mongo/MongoReactiveHealthIndicatorTest.java (hunk 2 above)
Comment (qinnnyul) [reply spring-boot-project/spring-boot-actuator/src/test/java/org/springframework/boot/actuate/mongo/MongoReactiveHealthIndicatorTest.java]: completed.

File: spring-boot-project/spring-boot-actuator-autoconfigure/pom.xml (hunk 3 above)
Comment (qinnnyul) [reply spring-boot-project/spring-boot-actuator-autoconfigure/pom.xml]: dependencies been re-ordered.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/mongo/MongoReactiveDataAutoConfiguration.java (hunk 1 above)
Comment (snicoll) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/mongo/MongoReactiveDataAutoConfiguration.java]: @qinnnyul sorry I am not  following. Why is it required?

Comment (qinnnyul) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/mongo/MongoReactiveDataAutoConfiguration.java]: @snicoll In current class, we can see “reactiveMongoTemplate”  method, when it constructs instance, it needs spring to inject two type of beans(one is ReactiveMongoDatabaseFactory,another is MongoConverter),  however, it only has method to construct ReactiveMongoDatabaseFactory before, another is missing. You can find the similar way in MongoDataAutoConfiguration.class, but I did not want to mix it up, that is why I added default implementation of MongoConverter here, does that make sense to you? what do you recommend?

Comment (snicoll) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/mongo/MongoReactiveDataAutoConfiguration.java]: I saw that but this is provided by `MongoDataAutoConfiguration` and there isn't a reason to copy/paste that here. Can you please rework the PR so that tests also include `MongoDataAutoConfiguration`?

Comment (snicoll) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/mongo/MongoReactiveDataAutoConfiguration.java]: See also #12001

Comment (qinnnyul) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/mongo/MongoReactiveDataAutoConfiguration.java]: @snicoll thanks for your suggestion, but I am not sure if it is good idea to make an explicit link between MongoReactiveDataAutoConfiguration and MongoDataAutoConfiguration, cause it has different conditions. Besides that, even I tried what you suggested, it throws out the following error: “Error creating bean with name 'reactiveMongoTemplate' defined in class path resource [org/springframework/boot/autoconfigure/data/mongo/MongoReactiveDataAutoConfiguration.class]: Unsatisfied dependency expressed through method 'reactiveMongoTemplate' parameter 1; nested exception is org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'org.springframework.data.mongodb.core.convert.MongoConverter' available: “

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/mongo/MongoReactiveDataAutoConfiguration.java [hunk 4]
```diff
@@ -43,12 +44,13 @@
...
  *
  * @author Mark Paluch
+ * @author Yulin Qin
  * @since 2.0.0
  */
 @Configuration
 @ConditionalOnClass({ MongoClient.class, ReactiveMongoTemplate.class })
 @EnableConfigurationProperties(MongoProperties.class)
-@AutoConfigureAfter(MongoReactiveAutoConfiguration.class)
+@AutoConfigureAfter({ MongoReactiveAutoConfiguration.class, MongoDataAutoConfiguration.class })
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/mongo/MongoReactiveDataAutoConfiguration.java]: Please rebase your PR rather than changing something that has already been changed.

File: spring-boot-project/spring-boot-actuator/src/test/java/org/springframework/boot/actuate/mongo/MongoReactiveHealthIndicatorTest.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+import static org.mockito.Mockito.mock;
+
+/**
+ * Tests for {@link MongoReactiveHealthIndicator}.
+ *
+ * @author Yulin Qin
+ */
+
+public class MongoReactiveHealthIndicatorTest {
+	private MongoReactiveHealthIndicator mongoReactiveHealthIndicator;
```
Comment (snicoll) [inline spring-boot-project/spring-boot-actuator/src/test/java/org/springframework/boot/actuate/mongo/MongoReactiveHealthIndicatorTest.java]: This doesn't need to be a field

File: spring-boot-project/spring-boot-actuator/src/test/java/org/springframework/boot/actuate/mongo/MongoReactiveHealthIndicatorTest.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.BDDMockito.given;
+import static org.mockito.Mockito.mock;
+
+/**
+ * Tests for {@link MongoReactiveHealthIndicator}.
+ *
+ * @author Yulin Qin
+ */
+
```
Comment (snicoll) [inline spring-boot-project/spring-boot-actuator/src/test/java/org/springframework/boot/actuate/mongo/MongoReactiveHealthIndicatorTest.java]: Unnecessary extra space.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/mongo/MongoReactiveHealthIndicatorAutoConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+ *
+ * @author Yulin Qin
+ * @since 2.0.0
+ */
+@Configuration
+@ConditionalOnClass(ReactiveMongoTemplate.class)
+@ConditionalOnBean(ReactiveMongoTemplate.class)
+@ConditionalOnEnabledHealthIndicator("mongo")
+@AutoConfigureBefore(HealthIndicatorAutoConfiguration.class)
+@AutoConfigureAfter({MongoReactiveAutoConfiguration.class, MongoReactiveDataAutoConfiguration.class})
```
Comment (snicoll) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/mongo/MongoReactiveHealthIndicatorAutoConfiguration.java]: I think only the latter is necessary here.

File: spring-boot-project/spring-boot-actuator-autoconfigure/pom.xml [hunk 8]
```diff
@@ -331,6 +331,16 @@
 			<artifactId>spring-session-core</artifactId>
 			<optional>true</optional>
 		</dependency>
+		<dependency>
+			<groupId>org.mongodb</groupId>
```
Comment (snicoll) [inline spring-boot-project/spring-boot-actuator-autoconfigure/pom.xml]: This isn't sorted properly still (m is before s)

Comment (qinnnyul) [reply spring-boot-project/spring-boot-actuator-autoconfigure/pom.xml]: done

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/mongo/MongoReactiveHealthIndicatorAutoConfiguration.java (hunk 7 above)
Comment (qinnnyul) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/mongo/MongoReactiveHealthIndicatorAutoConfiguration.java]: changed

File: spring-boot-project/spring-boot-actuator/src/test/java/org/springframework/boot/actuate/mongo/MongoReactiveHealthIndicatorTest.java (hunk 5 above)
Comment (qinnnyul) [reply spring-boot-project/spring-boot-actuator/src/test/java/org/springframework/boot/actuate/mongo/MongoReactiveHealthIndicatorTest.java]: done

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/mongo/MongoReactiveDataAutoConfiguration.java (hunk 4 above)
Comment (qinnnyul) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/mongo/MongoReactiveDataAutoConfiguration.java]: done

**Comment by snicoll** [comment]
@qinnnyul Thanks for working on my comments. Unfortunately, the state of this PR is still poor so I decided to [polish it myself](https://github.com/spring-projects/spring-boot/commit/28f5392787dba92354d652a21cdc62fe3565af4a) rather than asking you to review again. Please review the following and take it into account for any future PR:

* The PR wasn't properly rebased on master. You can run `git rebase origin/master` on your branch once you've pulled the latest change (`git fetch origin`). Some conflicts needed to be fixed. Once that's done, please squash your commit into one and push force on your branch
* The build did not pass. Did you run it? Some tests were failing because the reactive driver wasn't available. Once I added it, other tests start failing as they didn't expect reactive mongo to be there
* The reactive health indicator wasn't shadowing the "regular" health indicator so an app with the reactive mongo driver had two health indicators for MongoDB. You wrote a test that was supposed to exercise that but it wasn't importing the regular mongo auto-configuration. Once I added that, it failed
* reactive health indicators are listed in the doc

**Comment by qinnnyul** [comment]
@snicoll thanks, will keep it in mind.
